integration/oasis/stationfinder/stations_iterator_alg.go

```diff
@@ -129,3 +131,96 @@ func buildChargeStationInfoDict(iter nearbyStationsIterator) map[string]bool {
 
 	return dict
 }
+
+type WeightBetweenNeighbors struct {
+	c   []CostBetweenChargeStations
+	err error
+}
+
+// CalculateWeightBetweenNeighbors accepts locations array, which will search for nearby
+// charge stations and then calculate weight between stations, the result is used to
+// construct graph.
+// - The input of locations contains: orig location -> first place to search for charge ->
+//   second location to search for charge -> ... -> dest location
+// - Both search nearby charge stations and calculate weight between stations are heavy
+//   operations, so put them into go-routine and use waitgroup to guarantee result channel
+//   is closed after everything is done.
+// - CalcWeightBetweenChargeStationsPair needs two iterators, one for nearbystationiterator
+//   represents from location and one for next location.  An array of channel is created
+//   to represent whether specific iterator is ready or not.
+// - All iterators has been recorded in iterators array
+//   @Todo: isIteratorReady could be removed later.  When iterator is not ready, should
+//         pause inside iterator itself.  That need refactor the design of stationfinder.
+func CalculateWeightBetweenNeighbors(locations []*StationCoordinate, oc *osrmconnector.OSRMConnector, sc *searchconnector.TNSearchConnector) chan WeightBetweenNeighbors {
+	c := make(chan WeightBetweenNeighbors)
+
+	if len(locations) > 2 {
+		iterators := make([]nearbyStationsIterator, len(locations))
+		isIteratorReady := make([]chan bool, len(locations))
+		for i := range isIteratorReady {
+			isIteratorReady[i] = make(chan bool)
+		}
+		var wg sync.WaitGroup
+
+		for i := 0; i < len(locations); i++ {
+			if i == 0 {
+				wg.Add(1)
+				go func(wg *sync.WaitGroup, first int) {
+					iterators[first] = NewOrigIter(locations[first])
+					isIteratorReady[first] <- true
+					wg.Done()
+					glog.Info("Finish generating NewOrigIter")
+				}(&wg, i)
+				continue
+			}
+
+			if i == len(locations)-1 {
+				wg.Add(1)
+				go func(wg *sync.WaitGroup, last int) {
+					iterators[last] = NewDestIter(locations[last])
+					glog.Info("Finish generating NewDestIter")
+					<-isIteratorReady[last-1]
+					putWeightBetweenChargeStationsIntoChannel(iterators[last-1], iterators[last], c, oc)
+					glog.Infof("Finish generating putWeightBetweenChargeStationsIntoChannel for %d", last)
+					wg.Done()
+				}(&wg, i)
+
+				break
+			}
+
+			wg.Add(1)
+			go func(wg *sync.WaitGroup, index int) {
+				iterators[index] = NewLowEnergyLocationStationFinder(sc, locations[index])
+				glog.Infof("Finish generating NewLowEnergyLocationStationFinder for %d", index)
+				<-isIteratorReady[index-1]
+				isIteratorReady[index] <- true
+				putWeightBetweenChargeStationsIntoChannel(iterators[index-1], iterators[index], c, oc)
+				glog.Infof("Finish generating putWeightBetweenChargeStationsIntoChannel for %d", index)
+				wg.Done()
+			}(&wg, i)
+		}
+
+		go func(wg *sync.WaitGroup) {
```

Comment:
Is it necesary to have another goroutine to wait? Too many async functions may cause issues due to complexity. 
Some suggestions here: https://github.com/golang/go/wiki/CodeReviewComments#synchronous-functions

Reply:
Fair enough.